# Naive_Bayes_Classifier
Naive Bayes is among one of the simplest, but most powerful algorithms for classification based on Bayes' Theorem with an assumption of independence among predictors. The Naive Bayes model is easy to build and particularly useful for very large data sets.

In [0]:
from sklearn import datasets                          # Import scikit-learn dataset library
from sklearn import metrics                           # Import scikit-learn metrics library to calcute the prediction metrices.
from sklearn.model_selection import train_test_split  # import scikit-learn train_test_split library for pliting the dataset into train and test.
from sklearn import naive_bayes                       # Import scikit-learn naive_bayes library

**Here I am using prebuilt -  wine - dataset in sklearn.**

In [0]:
dataset = datasets.load_wine()  #Load dataset - wine.

In [0]:
print(dataset.data.shape)   # return the shape of the dataset.
print(dataset.target.shape) # return the size of the target column.

(178, 13)
(178,)


Thw wine dataset is consists of 178 rows and 12 features. 

In [0]:
print(dataset.feature_names)  # return the name of the columns in the dataset.
print('\n')
print(dataset.target_names)   # return the values in the target column.
print('\n')
print(dataset.target)         # print the target set values.

['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']


['class_0' 'class_1' 'class_2']


[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2]


Dataset comprises of 13 features (alcohol, malic_acid, ash, alcalinity_of_ash, magnesium, total_phenols, flavanoids, nonflavanoid_phenols, proanthocyanins, color_intensity, hue, od280/od315_of_diluted_wines, proline) and type of wine cultivar. This data has three type of wine Class_0, Class_1, and Class_3. Here you can build a model to classify the type of wine. 

class_0 : 0

class_1 : 1

class_2 : 2

In [0]:
X = dataset.data; y = dataset.target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)
print(X.shape)
print(X_train.shape)
print(X_test.shape)

(178, 13)
(133, 13)
(45, 13)


First, you separate the columns into dependent and independent variables(or features and label). Then you split those variables into train and test set. The set is splited in 75(train):25 (test) propotion. The test dataset is consists of total 45 rows. The target values are assigned to numpy array Y.

In [0]:
import pandas as pd                                             # imorting pands library
df = pd.DataFrame(dataset.data, columns=dataset.feature_names)  # convert the numpy array in to pandas dataframe
df.head()                                                       # print first 5 rows of the dataset.

alcohol  malic_acid   ash  ...   hue  od280/od315_of_diluted_wines  proline
0    14.23        1.71  2.43  ...  1.04                          3.92   1065.0
1    13.20        1.78  2.14  ...  1.05                          3.40   1050.0
2    13.16        2.36  2.67  ...  1.03                          3.17   1185.0
3    14.37        1.95  2.50  ...  0.86                          3.45   1480.0
4    13.24        2.59  2.87  ...  1.04                          2.93    735.0

[5 rows x 13 columns]

In [0]:
# fit a naive_bayes.BernoulliNB() model to the data

model = naive_bayes.BernoulliNB() #Create a Bernoulli Classifier
model.fit(X_train, y_train)       #Train the model using the training sets
print(); print(model)


BernoulliNB(alpha=1.0, binarize=0.0, class_prior=None, fit_prior=True)


In Bernoulli Naive Bayes, we assumes that all our features are binary such that they take only two values. Means 0s can represent “word does not occur in the document” and 1s as "word occurs in the document" .

In [0]:
# make predictions

expected_y  = y_test                    # storing the y_test value to expected_y.
predicted_y = model.predict(X_test)     # Predict the response for test dataset

In [0]:
# summarize the fit of the model

print(); print('naive_bayes.BernoulliNB(): ')
print(); print(metrics.classification_report(expected_y, predicted_y,target_names=dataset.target_names)) # return classification and confusion metrix.
print(); print(metrics.confusion_matrix(expected_y, predicted_y)) # return classification and confusion metrix.


naive_bayes.BernoulliNB(): 

              precision    recall  f1-score   support

     class_0       0.00      0.00      0.00        18
     class_1       0.29      1.00      0.45        13
     class_2       0.00      0.00      0.00        14

    accuracy                           0.29        45
   macro avg       0.10      0.33      0.15        45
weighted avg       0.08      0.29      0.13        45


[[ 0 18  0]
 [ 0 13  0]
 [ 0 14  0]]


In Bernoulli model, our accuracy is just 0.29, It predicted 18 wine types to class_0, 13 wine types belongs to class_1 and 14 wine entries to class_2.

To improve the accuracy of the prediction we will use: **naive_bayes.Gaussian model.**
and repeate the all codes.

In [0]:
# fit a naive_bayes.GaussianNB() model to the data

model = naive_bayes.GaussianNB()
model.fit(X_train, y_train)
print(); print(model)


GaussianNB(priors=None, var_smoothing=1e-09)


 Gaussian Naive Bayes is used in cases when all our features are continuous. Here features can have different values in data set. We can’t represent features in terms of their occurrences. This means data is continuous. Hence we use Gaussian Naive Bayes here.

In [0]:
# make predictions

expected_y  = y_test
predicted_y = model.predict(X_test)

In [0]:
# summarize the fit of the model

print(); print('naive_bayes.GaussianNB(): ')
print(); print(metrics.classification_report(expected_y, predicted_y, target_names=dataset.target_names))
print(); print(metrics.confusion_matrix(expected_y, predicted_y))


naive_bayes.GaussianNB(): 

              precision    recall  f1-score   support

     class_0       1.00      1.00      1.00        15
     class_1       1.00      1.00      1.00        19
     class_2       1.00      1.00      1.00        11

    accuracy                           1.00        45
   macro avg       1.00      1.00      1.00        45
weighted avg       1.00      1.00      1.00        45


[[15  0  0]
 [ 0 19  0]
 [ 0  0 11]]


**Accuracy of the model is 1.00 which is 100%.**